kvserver/rangefeed: add newRetryErrBufferCapacityExceeded

wenyihu6 opened the pull request:
This patch refactors the error for kvpb.RangeFeedRetryError_REASON_SLOW_CONSUMER
into newRetryErrBufferCapacityExceeded.

Part of: https://github.com/cockroachdb/cockroach/issues/126560
Release note: none

wenyihu6 committed:
to be squashed separated out for easier code review

wenyihu6 committed:
kvserver/rangefeed: add unbuffered registration tests

This patch adds more test cases for unbufferedRegistration.

Part of: #126560
Release note: none

wenyihu6 committed:
to be squashed separated out for easier code review

wenyihu6 committed:
kvserver/rangefeed: add newRetryErrBufferCapacityExceeded

This patch refactors the error for kvpb.RangeFeedRetryError_REASON_SLOW_CONSUMER
into newRetryErrBufferCapacityExceeded.

Part of: #126560
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/129232)
<!-- Reviewable:end -->

wenyihu6 committed:
kvserver/rangefeed: add rangefeed level callback

This patch adds RegisterRangefeedCleanUp to BufferedStream interface. Rangefeed
level can register a cleanup callback that will be invoked after
BufferedStream.Disconnect is called. Note that it might not be invoked
immediately during SendBufferedError.

Buffered registration has a dedicated goroutine to do cleanup. Unbuffered
registration does not have a dedicated goroutine, so it relies on BufferedSender
to invoke the callback when it disconnects.

Part of: #129814
Release note: none

wenyihu6 committed:
kvserver/rangefeed: add unbuffered registration

This patch adds unbufferedRegistration.

UnbufferedRegistration is like BufferedRangefeed but uses BufferedStream to
buffer live raft updates instead of a using buf channel and having a
dedicated per-range per-registration goroutine to volley events to underlying
grpc stream. Instead, there is only one BufferedStream for each incoming
node.MuxRangefeed rpc call. BufferedStream is responsible for buffering and
sending its updates to the underlying grpc stream in a dedicated goroutine
O(node).

Note that BufferedStreamSender is still left unimplemented. This commit doesn't
add any tests for it yet. More tests will be added in future commits of this PR.

Part of: #129814
Release note: none

wenyihu6 committed:
kvserver/rangefeed: move helper functions to registry_helpers_test

This patch moves helper functions to registry_helpers_test.

Part of: #129814
Release note: none

wenyihu6 committed:
kvserver/rangefeed: refactor registry_helpers_test to facilitate future tests

This patch refactors helper functions in registry_helpers_test.

Part of: #129814
Release note: none

wenyihu6 committed:
to be squashed separated out for easier code review

wenyihu6 committed:
to be squashed and discussed
